### PR TITLE
4900 backdated datetime

### LINKED ISSUE #4900: Backdate prescriptions - stock numbers are strange
## What went wrong? 😲

When you change the date on the prescription, then the stock numbers (available & in store) show values from the start of the date selected. This is problematic when the date is today, as any transactions made today aren't included.

This was compounded by an existing weirdness, which is that the 'in store' total is reduced by the quantity issued on the current prescription, but the 'available' is not. Thus you can have less stock in store than is available:

image

## Expected behaviour 🤔
Two changes would make things clearer:

* Change the time to 23:59 for the selected date
* Adjust the displayed value for the stock in store to be stock + quantity allocated to this prescription

The second item is how the available stock column behaves; we need to have both stock columns behave consistently

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Add some stock for item a
2. Create a prescription
3. Add item a
4. see that the stock which you've just added is not available: the full total shows in the 'in store' total but is not in the 'available' total

## Your environment 🌱

- Open mSupply Version:
- Legacy mSupply Central Server Version:

- Platform:

- Database type:

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #4900

# 👩🏻‍💻 What does this PR do?

This PR adds `backdated_datetime` to invoices.
The plan is to add this optional timestamp to mSupply as well.

Adding the extra field makes the purpose clearer (vs using allocated_datetime as the `backdated_datetime`).
This also makes it easier to update the other datetimes when the backdated date is changed...

## 💌 Any notes for the reviewer?

Is this change too big to go into 2.3.0? Is it safe, or do we need to re-test all the prescription stuff?

There is an edge case that if someone does manually change their prescription to as today, it will be treated as backdated, even though it's kind of going into the future (aka at midnight) I think that's ok?

# 🧪 Testing

- [ ] Create a normal prescription.
- [ ] Check that the appropriate date times are set e.g, `picked_datetime` and `created_datetime` are set correctly
- [ ] Create a backdated prescription
- [ ] Check that all the datetimes are set to midnight of the date backdated  too
- [ ] Try changing the backdated date at various stages make sure the dates are updated appropriately...

# 📃 Documentation
- [X] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour

### CODE REVIEW

File: client/packages/invoices/src/Prescriptions/api/operations.graphql [hunk 1]
```diff
@@ -2,7 +2,7 @@ fragment PrescriptionRow on InvoiceNode {
   __typename
   comment
   createdDatetime
-  prescriptionDate: allocatedDatetime
+  prescriptionDate: backdatedDatetime
```
Comment (jmbrunskill) [inline client/packages/invoices/src/Prescriptions/api/operations.graphql:5]: Kept this as prescriptionDate to minimise changes in frontend, I think it's ok how it is but maybe it's better to make it clearly backdatedDatetime in frontend?

File: server/service/src/invoice/prescription/insert/generate.rs [hunk 2]
```diff
@@ -40,7 +40,7 @@ pub fn generate(
         comment: None,
         their_reference: None,
         transport_reference: None,
-        allocated_datetime: Some(current_datetime),
+        allocated_datetime: None,
```
Comment (jmbrunskill) [inline server/service/src/invoice/prescription/insert/generate.rs:43]: Removed this default, so allocated_datetime won't be set now, unless backdating...

File: server/service/src/invoice/prescription/update/generate.rs [hunk 3]
```diff
@@ -27,28 +27,29 @@ pub(crate) fn generate(
...
         should_update_batches_total_number_of_packs(&existing_invoice, &input_status);
     let mut update_invoice = existing_invoice.clone();
 
-    // prescription_date is None if the invoice is not backdated, this means we get actual picked/verified date
-    // If we have a backdated invoice, we use the backdated date for picked/verified date
-    let prescription_date = match prescription_date {
-        Some(prescription_date) => Some(prescription_date),
+    let backdated_datetime = match backdated_datetime {
+        Some(backdated_datetime) => {
+            backdate_status_datetimes(&mut update_invoice, backdated_datetime);
```
Comment (jmbrunskill) [inline server/service/src/invoice/prescription/update/generate.rs:40]: Now always updating datetimes if we change the backdated time.

File: server/service/src/invoice/prescription/update/generate.rs [hunk 4]
```diff
@@ -95,10 +96,30 @@ fn should_update_batches_total_number_of_packs(
...
 }
 
+fn backdate_status_datetimes(invoice: &mut InvoiceRow, backdated_datetime: NaiveDateTime) {
+    match &invoice.status {
+        InvoiceStatus::New => {} // Created datetime is never backdated
+        InvoiceStatus::Allocated => {
+            invoice.allocated_datetime = Some(backdated_datetime);
+        }
+        InvoiceStatus::Picked => {
+            invoice.allocated_datetime = Some(backdated_datetime);
```
Comment (jmbrunskill) [inline server/service/src/invoice/prescription/update/generate.rs:106]: Is this safe, should we always have an allocated date? Maybe we should check if it's not null before setting it?

File: server/service/src/invoice_line/stock_out_line/mod.rs [hunk 5]
```diff
@@ -35,9 +35,9 @@ impl StockOutType {
 }
 
 pub(crate) fn invoice_backdated_date(invoice: &InvoiceRow) -> Option<NaiveDateTime> {
-    if let Some(allocated_date) = invoice.allocated_datetime.clone() {
-        if allocated_date < invoice.created_datetime {
-            return Some(allocated_date);
+    if let Some(backdated_datetime) = invoice.backdated_datetime.clone() {
+        if backdated_datetime < invoice.created_datetime {
```
Comment (jmbrunskill) [inline server/service/src/invoice_line/stock_out_line/mod.rs:39]: I think this logic is still good, if backdated time is after created_datetime, it's sort of considered not a backdated invoice, although the field is populated in the db which could cause some confusion in reporting (if that was to happen) e.g. # of backdated invoiced could be incorrect if people set it to today a lot...

File: client/packages/invoices/src/Prescriptions/api/operations.graphql (hunk 1 above)
Comment (andreievg) [reply client/packages/invoices/src/Prescriptions/api/operations.graphql:5]: Yeah I think backdateDatetime would be better for front end

File: server/service/src/sync/translations/invoice.rs [hunk 6]
```diff
@@ -181,6 +181,11 @@ pub struct LegacyTransactRow {
     #[serde(rename = "om_original_shipment_id")]
     #[serde(deserialize_with = "empty_str_as_option_string")]
     pub original_shipment_id: Option<String>,
+
+    #[serde(default)]
+    #[serde(rename = "backdated_datetime")]
```
Comment (andreievg) [inline server/service/src/sync/translations/invoice.rs:186]: I think shis would be om_backdated_datetime in mSupply

File: server/service/src/sync/translations/invoice.rs [hunk 7]
```diff
@@ -298,6 +303,14 @@ impl SyncTranslation for InvoiceTranslation {
             }
         };
 
+        // If it was created after allocation, then it must be a backdated invoice
+        let backdated_datetime = match mapping.created_datetime
```
Comment (andreievg) [inline server/service/src/sync/translations/invoice.rs:307]: Is this for backwards compatibility with RC build ? I don't think we need to support it

File: server/service/src/invoice_line/stock_out_line/mod.rs (hunk 5 above)
Comment (andreievg) [reply server/service/src/invoice_line/stock_out_line/mod.rs:39]: > There is an edge case that if someone does manually change their prescription to as today, it will be treated as backdated,

With this change that other edge case is dealt with ?

File: server/service/src/invoice/prescription/update/generate.rs (hunk 4 above)
Comment (andreievg) [reply server/service/src/invoice/prescription/update/generate.rs:106]: I find it easer to think about if we first do normal update of statuses, and then adjust them by backdate (for those dates that are set)

https://github.com/msupply-foundation/open-msupply/compare/4900-backdated_datetime...4900-backdated_datetime(PR-review-suggestion)

**Review by andreievg — CHANGES_REQUESTED** [review]
Thanks, have some small suggestions.

Also do you think we should be updating prescription datetime if you click in and out of prescription datetime, but it hasn't changed ? (like when by default it's current date, and we click in and out of it?)

File: server/service/src/sync/translations/invoice.rs (hunk 7 above)
Comment (jmbrunskill) [reply server/service/src/sync/translations/invoice.rs:307]: Was thinking for mSupply backdated, sounds like better to use picked_datetime from legacy...

File: server/service/src/sync/translations/invoice.rs (hunk 6 above)
Comment (jmbrunskill) [reply server/service/src/sync/translations/invoice.rs]: [2b93947](https://github.com/msupply-foundation/open-msupply/pull/4914/commits/2b9394723aac585574ed65df47b1477c7e62b5c7)

File: server/service/src/sync/translations/invoice.rs (hunk 7 above)
Comment (jmbrunskill) [reply server/service/src/sync/translations/invoice.rs]: [9124e6c](https://github.com/msupply-foundation/open-msupply/pull/4914/commits/9124e6c461955beee432d202b8ced76f2c01aa6f)

**Comment by lache-melvin** [comment]
@jmbrunskill was going to review, but I think develop got merged in here instead of v2.3.0 😁

**Comment by jmbrunskill** [comment]
@lache-melvin dang, that was me, was trying to resolve conflicts...
One tick...

File: server/service/src/invoice_line/stock_out_line/mod.rs (hunk 5 above)
Comment (jmbrunskill) [reply server/service/src/invoice_line/stock_out_line/mod.rs:39]: @andreievg  I think we agreed to leave this for now? I'll raise an issue for follow up though...

**Review by andreievg — APPROVED** [review]
Thanks @jmbrunskill, look great !

**Comment by jmbrunskill** [comment]
Also added https://github.com/msupply-foundation/open-msupply/issues/4980 for follow up...
